ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,316 @@
+    }
+    //can only be the case when (next_A_or_B == zigzag_waypoint.switch_pos_A)
+    // calculate next A
+    Vector2f pos_diff_AC = cur_pos_2d - zigzag_waypoint.A_pos;
+    if ((pos_diff_AC.x*pos_diff.x + pos_diff_AC.y*pos_diff.y) < 0) {
+        next_dir = -1;
+    }
+    float dis_CA_squared = (cur_pos_2d - zigzag_waypoint.A_pos).length_squared();
+    float dis_AE = sqrtf(dis_CA_squared - dis_from_AB_squared);
+    float dis_ratio = dis_AE / dis_AB;
+    next_dest.x = cur_pos.x - next_dir*dis_ratio*pos_diff.x;
+    next_dest.y = cur_pos.y - next_dir*dis_ratio*pos_diff.y;
+    return true;
+}
+
+// called by ZIGZAG case in RC_Channel.cpp
+// used to record point A, B and give the signal to fly to next destination automatically
+void Copter::ModeZigZag::receive_signal_from_auxsw(RC_Channel::aux_switch_pos_t aux_switch_position)
+{
+    // define point A and B
+    if (stage == REQUIRE_A || stage == REQUIRE_B) {
+        if (aux_switch_position != RC_Channel::aux_switch_pos_t::MIDDLE) {
+            Vector3f cur_pos = inertial_nav.get_position();
+            set_destination(cur_pos, aux_switch_position);
+            return;
+        }
+    } else {      
+        // A and B have been defined
+        if (aux_switch_position != RC_Channel::aux_switch_pos_t::MIDDLE) { //switch position in HIGH or LOW
+            // calculate next point A or B
+            // need to judge if the drone's position is between A and B
+            Vector3f next_dest;
+            if(calculate_next_dest(next_dest, aux_switch_position)){
+                // initialise waypoint and spline controller
+                wp_nav->wp_and_spline_init();
+                set_destination(next_dest, aux_switch_position);
+                // initialise yaw
+                auto_yaw.set_mode_to_default(false);
+                stage = AUTO;
+                zigzag_is_between_A_and_B = false;
+            }
+        } else {      //switch in middle position, regain the control
+            if (stage == AUTO) {
+                stage = MANUAL_REGAIN;
+                loiter_nav->init_target();
+                zigzag_is_between_A_and_B = true;
+            } else {
+                zigzag_is_between_A_and_B = false;
+            }
+        }
+    }
+}
+
+// set_destination - sets zigzag mode's target destination
+// Returns true if the fence is enabled and guided waypoint is within the fence
+// else return false if the waypoint is outside the fence
+bool Copter::ModeZigZag::set_destination(const Vector3f& destination, RC_Channel::aux_switch_pos_t aux_switch_position)
+{
+
+#if AC_FENCE == ENABLED
+    // reject destination if outside the fence
+    Location_Class dest_loc(destination);
+    if (!copter.fence.check_destination_within_fence(dest_loc)) {
+        copter.Log_Write_Error(ERROR_SUBSYSTEM_NAVIGATION, ERROR_CODE_DEST_OUTSIDE_FENCE);
+        // failure is propagated to GCS with NAK
+        return false;
+    }
+#endif
+    switch (stage) {
+        case REQUIRE_A:
```

Comment:
Indenting is incorrect; case should be under switch

Reply:
although personally I actually prefer the case to be indented..